### PR TITLE
 KFSPTS-12176 adding DVs to the create accounting document job

### PR DESCRIPTION
**Pull request by jhulslander**
This is ready for review.  Please note, in the unit tests, I added verifications for some variables in the sub objects of the Disbursement Voucher document.  I didn't add validations for each value as there a lot, and thought it was sufficient to just verify the sub objects had some values loaded.  If you think every value should be verified, please let me know, and will add the extra checks.

### CODE REVIEW

File: src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherPreConferenceRegistrant.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    public void setDepartmentCode(String departmentCode) {
+        this.departmentCode = departmentCode;
+    }
+
+    public String getPreConferenceRequestNuumber() {
+        return preConferenceRequestNuumber;
+    }
+
+    public void setPreConferenceRequestNuumber(String preConferenceRequestNuumber) {
+        this.preConferenceRequestNuumber = preConferenceRequestNuumber;
```
Comment (ajd299) [inline src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherPreConferenceRegistrant.java]: typo in method name, parameter and member variable preConferenceRequestNuumber

File: src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherNonEmployeeTravel.java [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+
+import org.apache.commons.lang3.StringUtils;
+import org.kuali.rice.core.api.util.type.KualiDecimal;
+
+import edu.cornell.kfs.sys.xmladapters.KualiDecimalNullPossibleXmlAdapter;
+import edu.cornell.kfs.sys.xmladapters.KualiDecimalXmlAdapter;
+import edu.cornell.kfs.sys.xmladapters.StringToJavaDateAdapter;
+
+@XmlAccessorType(XmlAccessType.FIELD)
+@XmlRootElement(name = "NonEmoployeeTravel", namespace = StringUtils.EMPTY)
```
Comment (ajd299) [inline src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherNonEmployeeTravel.java]: typo

File: src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherNonEmployeeTravel.java [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+public class DisbursementVoucherNonEmployeeTravel {
+    
+    @XmlElement(name = "travler_name", namespace = StringUtils.EMPTY, required = false)
```
Comment (ajd299) [inline src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherNonEmployeeTravel.java]: typo

File: src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherNonEmployeeExpense.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+import javax.xml.bind.annotation.adapters.NormalizedStringAdapter;
+import javax.xml.bind.annotation.adapters.XmlJavaTypeAdapter;
+
+import org.apache.commons.lang3.StringUtils;
+import org.kuali.rice.core.api.util.type.KualiDecimal;
+
+import edu.cornell.kfs.sys.xmladapters.KualiDecimalXmlAdapter;
+
+@XmlAccessorType(XmlAccessType.FIELD)
+@XmlRootElement(name = "NonEmoployeeTravelExpense", namespace = StringUtils.EMPTY)
```
Comment (ajd299) [inline src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherNonEmployeeExpense.java]: typo

File: src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherNonEmployeeTravel.java [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+    protected KualiDecimal perdiemActualAmount;
+    
+    @XmlElement(name = "round_trip_code", namespace = StringUtils.EMPTY, required = false)
+    @XmlJavaTypeAdapter(NormalizedStringAdapter.class)
+    protected String roundTripCode;
+    
+    @XmlElement(name = "personal_car_milage_amount", namespace = StringUtils.EMPTY, required = false)
+    @XmlJavaTypeAdapter(KualiDecimalXmlAdapter.class)
+    protected KualiDecimal personalCarMilageAmount;
+
```
Comment (ajd299) [inline src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherNonEmployeeTravel.java]: typos above (milage => mileage)

File: src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherNonEmployeeTravel.java [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+    
+    @XmlElement(name = "perdiem_change_reason_text", namespace = StringUtils.EMPTY, required = false)
+    @XmlJavaTypeAdapter(NormalizedStringAdapter.class)
+    protected String perdiemChangeReasonText;
+    
+    @XmlElement(name = "personal_car_amount", namespace = StringUtils.EMPTY, required = false)
+    @XmlJavaTypeAdapter(KualiDecimalNullPossibleXmlAdapter.class)
+    protected KualiDecimal personalCarAmount;
+    
+    @XmlElementWrapper(name = "travler_expenses", namespace = StringUtils.EMPTY, required = true)
```
Comment (ajd299) [inline src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherNonEmployeeTravel.java]: typo above (travler => traveler)?

File: src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherNonEmployeeTravel.java [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+    }
+
+    public KualiDecimal getPersonalCarMilageAmount() {
+        return personalCarMilageAmount;
+    }
+
+    public void setPersonalCarMilageAmount(KualiDecimal personalCarMilageAmount) {
+        this.personalCarMilageAmount = personalCarMilageAmount;
+    }
+
```
Comment (ajd299) [inline src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherNonEmployeeTravel.java]: typos in the getter/setter (Milage => Mileage)

File: src/main/java/edu/cornell/kfs/fp/batch/service/impl/CuDisbursementVoucherDocumentGenerator.java [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+import org.kuali.kfs.krad.bo.Note;
+import org.kuali.kfs.krad.util.ObjectUtils;
+import org.kuali.kfs.sys.businessobject.AccountingLine;
+import org.kuali.kfs.sys.service.UniversityDateService;
+import org.kuali.rice.core.api.util.type.KualiDecimal;
+
+import edu.cornell.kfs.fp.batch.xml.AccountingXmlDocumentAccountingLine;
+import edu.cornell.kfs.fp.batch.xml.AccountingXmlDocumentEntry;
+import edu.cornell.kfs.fp.batch.xml.DisbursementVoucherDetail;
+import edu.cornell.kfs.fp.batch.xml.DisbursementVoucherNonEmployeeExpense;
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/fp/batch/service/impl/CuDisbursementVoucherDocumentGenerator.java]: I think adding xml to the end of these classes would make sense.  pre-pending the classes with CU would imply they are Cornell extension of the full Kuali business object, so I don't think we want to do that.  I will create a new user story to rename the preexisting classes edu.cornell.kfs.fp.batch.xml to include an XML at the end, where appropriate.

**Comment by jhulslander** [comment]
Thanks for catching those errors,  I believe I have fixed all of the catches.  I also added additional unit tests to verify the names changes to the XML marshaling objects didn't break anything.

File: src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherPaymentInfomrationXml.java [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+import javax.xml.bind.annotation.adapters.XmlJavaTypeAdapter;
+
+import org.apache.commons.lang3.StringUtils;
+import org.kuali.rice.core.api.util.type.KualiDecimal;
+
+import edu.cornell.kfs.sys.xmladapters.KualiDecimalXmlAdapter;
+import edu.cornell.kfs.sys.xmladapters.StringToJavaDateAdapter;
+
+@XmlAccessorType(XmlAccessType.FIELD)
+@XmlRootElement(name = "PaymentInformation", namespace = StringUtils.EMPTY)
```
Comment (ajd299) [inline src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherPaymentInfomrationXml.java]: Should this be payment_information?

File: src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherPreConferenceRegistrantXml.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    @XmlJavaTypeAdapter(NormalizedStringAdapter.class)
+    protected String name;
+    
+    @XmlElement(name = "department_code", namespace = StringUtils.EMPTY, required = false)
+    @XmlJavaTypeAdapter(NormalizedStringAdapter.class)
+    protected String departmentCode;
+    
+    @XmlElement(name = "pre_conference_request_number", namespace = StringUtils.EMPTY, required = false)
+    @XmlJavaTypeAdapter(NormalizedStringAdapter.class)
+    protected String preConferenceRequestNuumber;
```
Comment (ajd299) [inline src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherPreConferenceRegistrantXml.java]: Typo Nuumber => Number. Also, the Getter and setter names/variables can be updated.

File: src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherPreConferenceRegistrantXml.java [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+import javax.xml.bind.annotation.adapters.NormalizedStringAdapter;
+import javax.xml.bind.annotation.adapters.XmlJavaTypeAdapter;
+
+import org.apache.commons.lang3.StringUtils;
+import org.kuali.rice.core.api.util.type.KualiDecimal;
+
+import edu.cornell.kfs.sys.xmladapters.KualiDecimalXmlAdapter;
+
+@XmlAccessorType(XmlAccessType.FIELD)
+@XmlRootElement(name = "DisbursementVoucherPreConferenceRegistrant", namespace = StringUtils.EMPTY)
```
Comment (ajd299) [inline src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherPreConferenceRegistrantXml.java]: Should the name be registrant to match the documentation?

File: src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherDetailXml.java [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+import javax.xml.bind.annotation.XmlAccessorType;
+import javax.xml.bind.annotation.XmlElement;
+import javax.xml.bind.annotation.XmlRootElement;
+import javax.xml.bind.annotation.adapters.NormalizedStringAdapter;
+import javax.xml.bind.annotation.adapters.XmlJavaTypeAdapter;
+
+import org.apache.commons.lang3.StringUtils;
+
+@XmlAccessorType(XmlAccessType.FIELD)
+@XmlRootElement(name = "DisbursementVoucherDetail", namespace = StringUtils.EMPTY)
```
Comment (ajd299) [inline src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherDetailXml.java]: Does this correspond to dv_detail on the documentation?

File: src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherPrePaidTravelOverviewXml.java [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+import javax.xml.bind.annotation.XmlRootElement;
+import javax.xml.bind.annotation.adapters.NormalizedStringAdapter;
+import javax.xml.bind.annotation.adapters.XmlJavaTypeAdapter;
+
+import org.apache.commons.lang3.StringUtils;
+
+import edu.cornell.kfs.sys.xmladapters.StringToJavaDateAdapter;
+
+@XmlAccessorType(XmlAccessType.FIELD)
+@XmlRootElement(name = "DisbursementVoucherPrePaidTravelOverview", namespace = StringUtils.EMPTY)
```
Comment (ajd299) [inline src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherPrePaidTravelOverviewXml.java]: Should this be pre_paid_travel?

File: src/test/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImplTest.java [hunk 14]
```diff
@@ -498,6 +508,35 @@ private void assertNoteIsCorrect(Note expectedNote, Note actualNote) {
...
+        assertEquals("Wrong contact name", expectedDvDocument.getDisbVchrContactPersonName(), actualDvDocument.getDisbVchrContactPersonName());
+        assertEquals("payment reason code not correct", expectedDvDocument.getDvPayeeDetail().getDisbVchrPaymentReasonCode(), 
+                actualDvDocument.getDvPayeeDetail().getDisbVchrPaymentReasonCode());
+        assertEquals("payee type code not correct", expectedDvDocument.getDvPayeeDetail().getDisbursementVoucherPayeeTypeCode(), 
+                actualDvDocument.getDvPayeeDetail().getDisbursementVoucherPayeeTypeCode());
+        assertEquals("non employee travel perdiem rate not correct", expectedDvDocument.getDvNonEmployeeTravel().getDisbVchrPerdiemRate(),
+                actualDvDocument.getDvNonEmployeeTravel().getDisbVchrPerdiemRate());
+        assertEquals("conference destination not correct", expectedDvDocument.getDvPreConferenceDetail().getDvConferenceDestinationName(),
+                actualDvDocument.getDvPreConferenceDetail().getDvConferenceDestinationName());
+        assertEquals("non employee travler name not correct", expectedDvDocument.getDvNonEmployeeTravel().getDisbVchrNonEmpTravelerName(),
```
Comment (ajd299) [inline src/test/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImplTest.java]: Typo travler => traveler

File: src/test/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImplTest.java [hunk 15]
```diff
@@ -498,6 +508,35 @@ private void assertNoteIsCorrect(Note expectedNote, Note actualNote) {
...
+                actualDvDocument.getDvNonEmployeeTravel().getDvPersonalCarMileageAmount());
+        assertObjectListIsCorrect("non employee travel expenses aren't correct", expectedDvDocument.getDvNonEmployeeTravel().getDvNonEmployeeExpenses(), 
+                actualDvDocument.getDvNonEmployeeTravel().getDvNonEmployeeExpenses(), this::assertTravelExpenseCorrect);
+        assertObjectListIsCorrect("non employee prepaid travel expenses aren't correct", expectedDvDocument.getDvNonEmployeeTravel().getDvPrePaidEmployeeExpenses(), 
+                actualDvDocument.getDvNonEmployeeTravel().getDvPrePaidEmployeeExpenses(), this::assertTravelExpenseCorrect);
+    }
+    
+    private void assertTravelExpenseCorrect(DisbursementVoucherNonEmployeeExpense expectedExpense, DisbursementVoucherNonEmployeeExpense actualExpense) {
+        assertEquals("expense code not correct", expectedExpense.getDisbVchrExpenseCode(), actualExpense.getDisbVchrExpenseCode());
+        assertEquals("expense compnay name not correct", expectedExpense.getDisbVchrExpenseCompanyName(), actualExpense.getDisbVchrExpenseCompanyName());
```
Comment (ajd299) [inline src/test/java/edu/cornell/kfs/fp/batch/service/impl/CreateAccountingDocumentServiceImplTest.java]: typo compnay => company

File: src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/CuDisbursementVoucherDocumentFixture.java [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+    public final String contactName;
+    public final String paymentReasonCode;
+    public final String payeeTypeCode;
+    public final KualiDecimal perdiemRate;
+    public final String conferenceDestination;
+    public final String nonEmployeeTravelerName;
+    public final Integer nonEmployeeCarMileage;
+    public final List<CuDisbursementVoucherDocumentNonEmployeeTravelExpenseFixture> nonEmployeeTrevelerExpense;
+    public final List<CuDisbursementVoucherDocumentNonEmployeeTravelExpenseFixture> nonEmployeeTrevelerPrepaidExpense;
+
```
Comment (ajd299) [inline src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/CuDisbursementVoucherDocumentFixture.java]: The 2 variable names have a misspelling of Traveler (Treveler)

File: src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/CuDisbursementVoucherDocumentFixture.java [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+        this.conferenceDestination = StringUtils.EMPTY;
+        this.nonEmployeeTravelerName = StringUtils.EMPTY;
+        this.nonEmployeeCarMileage = new Integer(0);
+        this.nonEmployeeTrevelerExpense = new ArrayList<CuDisbursementVoucherDocumentNonEmployeeTravelExpenseFixture>();
+        this.nonEmployeeTrevelerPrepaidExpense = new ArrayList<CuDisbursementVoucherDocumentNonEmployeeTravelExpenseFixture>();
+    }
+    
+    private CuDisbursementVoucherDocumentFixture(String bankCode, String contactName, String paymentReasonCode, String payeeTypeCode, double perdiemRate, String conferenceDestination, 
+            String nonEmployeeTravelerName, Integer nonEmployeeCarMileage, CuDisbursementVoucherDocumentNonEmployeeTravelExpenseFixture[] nonEmployeeTrevelerExpenseArray,
+            CuDisbursementVoucherDocumentNonEmployeeTravelExpenseFixture[] nonEmployeeTrevelerPrepaidExpenseArray) {
```
Comment (ajd299) [inline src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/CuDisbursementVoucherDocumentFixture.java]: these 2 parameter names also have a misspelling (Treveler => Traveler)

File: src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/CuDisbursementVoucherDocumentNonEmployeeTravelExpenseFixture.java [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+
+public enum CuDisbursementVoucherDocumentNonEmployeeTravelExpenseFixture {
+    DELTA("A", "DELTA", 55.55),
+    OTHER_LODGING("L", "Other - Lodging", 1555.55),
+    PREPAID_AVIS("PR", "AVIS", 23.55),
+    PREPAID_OTHER("PO", "PREPAID OTHER", 230.55);
+    
+    
+    public final String expenseType;
+    public final String compnayName;
```
Comment (ajd299) [inline src/test/java/edu/cornell/kfs/fp/batch/xml/fixture/CuDisbursementVoucherDocumentNonEmployeeTravelExpenseFixture.java]: typo here compnay => company

File: src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherNonEmployeeExpenseXml.java [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+import javax.xml.bind.annotation.adapters.NormalizedStringAdapter;
+import javax.xml.bind.annotation.adapters.XmlJavaTypeAdapter;
+
+import org.apache.commons.lang3.StringUtils;
+import org.kuali.rice.core.api.util.type.KualiDecimal;
+
+import edu.cornell.kfs.sys.xmladapters.KualiDecimalXmlAdapter;
+
+@XmlAccessorType(XmlAccessType.FIELD)
+@XmlRootElement(name = "NonEmployeeTravelExpense", namespace = StringUtils.EMPTY)
```
[the remark this answers is not in the document]
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherNonEmployeeExpenseXml.java]: This corresponds to the expenses in traveler_expenses and prepaid_expenses, the root element annotation  in this set up doesn't need to in the XML being marshalled, just a uniquer annotation identifier for  jaxB

[the remark this answers is not in the document]
Comment (ajd299) [reply src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherNonEmployeeExpenseXml.java]: okay thanks for the explanation, please disregard my similar comments

File: src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherPaymentInfomrationXml.java (hunk 9 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherPaymentInfomrationXml.java]: no, that tag is on DisbursementVoucherDetailXml, where this attribute is used, this root element just needs to be unique.  I will go through the new ojbects, and check for consistency

File: src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherPreConferenceRegistrantXml.java (hunk 11 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherPreConferenceRegistrantXml.java]: The annotations of prepaid_registrants and registrant on DisbursementVoucherPrePaidTravelOverviewXml do match the documentation as fas as I can tell, the root element here just needs to be a unique identifier

File: src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherDetailXml.java (hunk 12 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherDetailXml.java]: Yes, and the dv_detail is in the annotation on AccountingXmlDocumentEntry

File: src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherPrePaidTravelOverviewXml.java (hunk 13 above)
Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/fp/batch/xml/DisbursementVoucherPrePaidTravelOverviewXml.java]: nope, this just needs to be unique, pre_paid_travel is annotated in DisbursementVoucherDetailXml

**Comment by jhulslander** [comment]
Thanks for noticing the root element tags, in the other XML objects, they have the same value as the annotation in where they are used, so to be consistent, I have changed these as well.

**Comment by jhulslander** [comment]
OK, local validation  looks good with the latest changes, this ready for review agian

**Comment by jhulslander** [comment]
I am going to close this PR, there a lot of comments that been fixed, I also just noticed some error handing isn't helpful

**Pull request closed without merging**